packages/flutter_tools/lib/src/build_system/targets/desktop.dart

```diff
@@ -7,59 +7,64 @@ import 'package:meta/meta.dart';
 Depfile unpackDesktopArtifacts({
   @required FileSystem fileSystem,
   @required List<String> artifacts,
   @required Directory outputDirectory,
-  @required String artifactPath,
+  @required String engineSourcePath,
+  @required String clientSourcePath,
 }) {
   final List<File> inputs = <File>[];
   final List<File> outputs = <File>[];
   for (final String artifact in artifacts) {
-    final String entityPath = fileSystem.path.join(artifactPath, artifact);
+    final String entityPath = fileSystem.path.join(engineSourcePath, artifact);
     final FileSystemEntityType entityType = fileSystem.typeSync(entityPath);
 
     if (entityType == FileSystemEntityType.notFound
+     || entityType == FileSystemEntityType.directory
      || entityType == FileSystemEntityType.link) {
       throw Exception('Unsupported file type: $entityType');
     }
-
-    // If this artifact is a file then copy the source over.
-    if (entityType == FileSystemEntityType.file) {
-      final String outputPath = fileSystem.path.join(
-        outputDirectory.path,
-        fileSystem.path.relative(entityPath, from: artifactPath),
-      );
-      final File destinationFile = fileSystem.file(outputPath);
-      if (!destinationFile.parent.existsSync()) {
-        destinationFile.parent.createSync(recursive: true);
-      }
-      final File inputFile = fileSystem.file(entityPath);
-      inputFile.copySync(destinationFile.path);
-      inputs.add(inputFile);
-      outputs.add(destinationFile);
-      continue;
+    assert(entityType == FileSystemEntityType.file);
+    final String outputPath = fileSystem.path.join(
+      outputDirectory.path,
+      fileSystem.path.relative(entityPath, from: engineSourcePath),
+    );
+    final File destinationFile = fileSystem.file(outputPath);
+    if (!destinationFile.parent.existsSync()) {
+      destinationFile.parent.createSync(recursive: true);
     }
-
-    // If the artifact is a directory, recursively copy every file from it.
-    for (final File input in fileSystem.directory(entityPath)
-      .listSync(recursive: true)
-      .whereType<File>()) {
-      final String outputPath = fileSystem.path.join(
-        outputDirectory.path,
-        fileSystem.path.relative(input.path, from: artifactPath),
-      );
-      final File destinationFile = fileSystem.file(outputPath);
-      if (!destinationFile.parent.existsSync()) {
-        destinationFile.parent.createSync(recursive: true);
-      }
-      final File inputFile = fileSystem.file(input);
-      inputFile.copySync(destinationFile.path);
-      inputs.add(inputFile);
-      outputs.add(destinationFile);
+    final File inputFile = fileSystem.file(entityPath);
+    inputFile.copySync(destinationFile.path);
+    inputs.add(inputFile);
+    outputs.add(destinationFile);
+  }
+  final Directory clientSourceDirectory = fileSystem.directory(clientSourcePath);
+  if (!clientSourceDirectory.existsSync()) {
+    throw Exception('Missing clientSourceDirectory: $clientSourcePath');
```

Comment:
GTK won't have this (GTK uses a C API for everything, so we'll be consistent with that instead of having a wrapper), so we shouldn't hardwire in the idea that it's required.

Reply:
Interesting! in that case I can make this optional and only check if it is provided

Reply:
Done